quelpa.el

```diff
@@ -1779,15 +1819,37 @@ the `quelpa' command has been run in the current Emacs session."
       (when quelpa-self-upgrade-p
         (quelpa-self-upgrade))
       (setq quelpa-cache
-            (cl-remove-if (lambda (package)
-                            (or (eq package 'quelpa)
-                                (not (package-installed-p package))))
-                          quelpa-cache :key #'car))
+            (cl-remove-if-not (lambda (item) (package-installed-p (car item))) quelpa-cache))
       (mapc (lambda (item)
               (when (package-installed-p (car (quelpa-arg-rcp item)))
                 (quelpa item)))
             quelpa-cache))))
 
+;;;###autoload
+(defun quelpa-upgrade (rcp &optional action)
+  "Upgrade a package found in `quelpa-cache' with recipe RCP.
+Optionally, ACTION can be passed for non-interactive call with value of:
+- `force' (or \\[universal-argument] \\[quelpa-upgrade]) for forced upgrade.
+- `current' (or \\[universal-argument] \\[universal-argument] \\[quelpa-upgrade])
+  for upgrade using current working tree."
+  (interactive
+   (when (quelpa-setup-p)
+     (let* ((quelpa-melpa-recipe-stores (list quelpa-cache))
+            (name (quelpa-interactive-candidate))
+            (prefix (prefix-numeric-value current-prefix-arg)))
+       (list (assoc name quelpa-cache)
+             (cond  ((eq prefix 4) 'force)
+                    ((eq prefix 16) 'current))))))
+  (when rcp
```

Comment:
Forgive me, but this logic seems very confusing, and I think it may indicate a need to refactor something, either here or elsewhere.

It seems to me that, to upgrade a package, the steps to follow should be approximately this:

1.  Delete old package version.
2.  Install new package version.

Each step should be idempotent.

It feels wrong to be binding upgrade-p variables and passing force-upgrade-p arguments to other functions.  Setting *one* of them may be necessary for controlling whether uncommitted changes are discarded, but other than that, the rest of the code shouldn't care whether a package is being installed or upgraded.

Also, I'm not sure we should *ever* discard uncommitted changes.  I tend to think that, if a user is messing with the local clones made by Quelpa, he ought to know what he's doing, and we should simply signal an error (perhaps printing the path to the clone) and let the user handle it.

What do you think?  Thanks.

Reply:
> It feels wrong to be binding upgrade-p variables and passing force-upgrade-p arguments to other functions. Setting _one_ of them may be necessary for controlling whether uncommitted changes are discarded, but other than that, the rest of the code shouldn't care whether a package is being installed or upgraded.

I guess the confusion is stemmed from `(quelpa rcp :force quelpa-force-upgrade-p)`.
The `:force quelpa-force-upgrade-p` can be remove, we've already bind `quelpa-force-upgrade-p` for that.
I'm not really sure about what `upgrade-p` is doing. But I guess it's there to indicate whether we want to upgrade the package or not. So we have to set it for the upgrade to works. @steckerhalter  for confirmation.

> It seems to me that, to upgrade a package, the steps to follow should be approximately this:
> 
> 1. Delete old package version.
> 2. Install new package version.
> 
> Each step should be idempotent.

To upgrade a package we need to:
1. Install new package version
2. Delete old package version

For now, Quelpa is only doing (1), the (2) is not yet done. We can probably try to do that in the future. User has to manually delete obsoleted packages for now.

> Also, I'm not sure we should _ever_ discard uncommitted changes. I tend to think that, if a user is messing with the local clones made by Quelpa, he ought to know what he's doing, and we should simply signal an error (perhaps printing the path to the clone) and let the user handle it.

Yes, with my change, by default Quelpa will notify user about that. However, user can choose to set `quelpa-force-upgrade-p` and overriding that behavior.

Reply:
> For now, Quelpa is only doing (1), the (2) is not yet done. We can probably try to do that in the future. User has to manually delete obsoleted packages for now.

yeah and if we do that in the future (which would be nice because to do it manually is a bit annoying) it would be better to delete obsolete packages afterwards because we don't know beforehand if the install fails and if it does we have successfully removed the package :smiley:

Reply:
I feel like discussion about this issue is getting a bit fragmented here.  Perhaps this is my mistake for putting comments that are more design-oriented into these code-oriented review comments.  Please see my comment at https://github.com/quelpa/quelpa/pull/152#discussion_r360722977 for how I think this should generally work.  I will mark this thread as resolved so we can discuss it elsewhere.